src/modules/keyboardmanager/KeyboardManagerEngineLibrary/KeyboardEventHandlers.cpp

```diff
@@ -579,46 +579,90 @@ namespace KeyboardEventHandlers
+                                    keyEventList = new INPUT[key_count]();
+                                    memset(keyEventList, 0, sizeof(keyEventList));
+                                    int i = 0;
+                                    Helpers::SetModifierKeyEvents(from, it->second.winKeyInvoked, keyEventList, i, false, KeyboardManagerConstants::KEYBOARDMANAGER_SHORTCUT_FLAG);
+                                    Helpers::SetKeyEvent(keyEventList, i, INPUT_KEYBOARD, (WORD)to, 0, KeyboardManagerConstants::KEYBOARDMANAGER_SHORTCUT_FLAG);
+                                }else
+                                {
+                                    Shortcut to = std::get<Shortcut>(newRemapping.targetShortcut);
+                                    key_count = from.Size() - 1 + to.Size() - 1 - 2* from.GetCommonModifiersCount(to) + 1;
+                                    keyEventList = new INPUT[key_count]();
+                                    memset(keyEventList, 0, sizeof(keyEventList));
 
-                            // Key up for all new shortcut keys, key down for original shortcut modifiers and current key press but common keys aren't repeated
-                            key_count = (dest_size) + (src_size - 1) - (2 * (size_t)commonKeys);
+                                    int i = 0;
+                                    Helpers::SetModifierKeyEvents(from, it->second.winKeyInvoked, keyEventList, i, false, KeyboardManagerConstants::KEYBOARDMANAGER_SHORTCUT_FLAG, to);
+                                    Helpers::SetModifierKeyEvents(to, it->second.winKeyInvoked, keyEventList, i, true, KeyboardManagerConstants::KEYBOARDMANAGER_SHORTCUT_FLAG, from);
 
-                            // If the target shortcut's action key is pressed, then it should be released and original shortcut's action key should be set
-                            bool isActionKeyPressed = false;
-                            if (ii.GetVirtualKeyState((std::get<Shortcut>(it->second.targetShortcut).GetActionKey())))
-                            {
-                                isActionKeyPressed = true;
-                                key_count += 2;
+                                    Helpers::SetKeyEvent(keyEventList, i, INPUT_KEYBOARD, (WORD)to.actionKey, 0, KeyboardManagerConstants::KEYBOARDMANAGER_SHORTCUT_FLAG);
+                                    newRemapping.isShortcutInvoked = true;
+                                }
+
+                                // Remember which win key was pressed initially
+                                if (ii.GetVirtualKeyState(VK_RWIN))
+                                {
+                                    newRemapping.winKeyInvoked = ModifierKey::Right;
+                                }
+                                else if (ii.GetVirtualKeyState(VK_LWIN))
+                                {
+                                    newRemapping.winKeyInvoked = ModifierKey::Left;
+                                }
```

Comment:
Why are the Win Keys special cases?

Reply:
It is used, at least, for remapping `Win + ActionKey -> ModifierKey + ActionKey`. The `Win` key doesn't have its own virtual code so it is remembered to restore the correct `Win` key. 


One can think about it in the following way. We need to maintain invariant for active remapping. And this part of the code makes `newRemapping` active so we need to set `winKeyInvoked`
https://github.com/microsoft/PowerToys/blob/4a1e21ac831eabcb0f0762d402677303b183df01/src/modules/keyboardmanager/KeyboardManagerEngineLibrary/KeyboardEventHandlers.cpp#L218-L226


I was tempted to delete it to simplify code but in this case, we should delete the win key from the following select list. Also, we would need more investigation here.

![image](https://user-images.githubusercontent.com/17161067/132021114-781648eb-3d57-4b83-82a3-e674fc380fe2.png)